[Kernel][Triton][FP8] Adding fp8 and variable length sequence support to Triton FAv2 kernel

rasmith opened the pull request:
This PR adds fp8 and variable length sequence support to Triton FAv2 kernel.

This kernel supports 8-bit KV cache, and also the following (forward only):

1) Fwd with causal masking
2) Arbitrary Q and KV sequence lengths
3) Arbitrary head sizes
4) Multi and grouped query attention
5) Variable sequence lengths
6) ALiBi and matrix bias
7) Supports fp8 for models, currently for Llama-3.1-8B-Instruct-FP8-QKV-Prob

This kernel is slightly faster than the current one:

Benchmarks:

================================
Llama-3-8B-Instruct

VLLM_USE_TRITON_FLASH_ATTN=1 python benchmarks/benchmark_latency.py --enable-chunked-prefill False --load-format dummy --batch-size 64 --num-iters-warmup 2 --num-iters 5 --input-len 2048 --output-len 128  --model /models/Llama-3-8B-Instruct/

old kernel:
Avg latency: 7.028674733600928 seconds

new kernel:
Avg latency: 6.267468033730983 seconds

===============================
Phi-3-medium-128k-instruct-quantized.w8a8

VLLM_USE_TRITON_FLASH_ATTN=1 python benchmarks/benchmark_latency.py --enable-chunked-prefill False --load-format dummy --batch-size 64 --num-iters-warmup 2 --num-iters 5 --input-len 2048 --output-len 128   --model /models/Phi-3-medium-128k-instruct-quantized.w8a8/

old kernel:
Avg latency: 10.266006066184492 seconds

new kernel:
Avg latency: 9.983664013911039 seconds

PPL Measurements:

context-size=1024 
sample-size=512 
max-model-len 32768

model=Llama-3.1-8B-Instruct-FP8-QKV-Prob

PPL=6.9046958999710615

model=Llama-3.1-8B-Instruct

PPL=6.5381874291070545

LM EVAL:

I included lm eval for KV and QKV-Prob models since the Prob model has the output of softmax(QK^T) quantized to fp8 before the second dot, which does seem to reduce accuracy.  Well, QKV-Prob probably has the most quantization, so lowest accuracy, I think.

LM EVAL: Llama-3.1-8B-Instruct-FP8-KV
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.796|±  |0.0255|
|     |       |strict-match    |     5|exact_match|↑  |0.732|±  |0.0281|


LM EVAL:  Llama-3.1-8B-Instruct-FP8-QKV-Prob 
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.756|±  |0.0272|
|     |       |strict-match    |     5|exact_match|↑  |0.592|±  |0.0311|

LM EVAL: Llama-3.1-8B-Instruct
 |Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.780|±  |0.0263|
|     |       |strict-match    |     5|exact_match|↑  |0.724|±  |0.0283|

rasmith committed:
Add kernel that supports variable sequence length

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
isort

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
codespell

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff/codespell

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
codespell

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

rasmith committed:
yapf

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

SageMoore commented:
Very Exciting! I'll give it a try tomorrow morning. Thanks for the kernel!

SageMoore commented:
Hi, @rasmith. Are you planning to add block_table support to this kernel? If so, great! Feel free to either add that here or make a follow on PR. If not, I'd be happy to look into this, but we will need it in order to integrate this kernel into V1.

maleksan85 commented:
> Hi, @rasmith. Are you planning to add block_table support to this kernel? If so, great! Feel free to either add that here or make a follow on PR. If not, I'd be happy to look into this, but we will need it in order to integrate this kernel into V1.

Hi @SageMoore, please try this if you need cache/block support: https://github.com/vllm-project/vllm/blob/e3f7ff65e7a6c08cd354f7f333bce543a4f0607e/vllm/attention/ops/prefix_prefill.py#L20

SageMoore commented:
I think there are a couple of things getting lost in communication here.

1. In order for this kernel to be used in vllm going forward, it needs to support mixing prefills and decodes into the same batch.
2. There's no need for a backwards pass kernel.
3. I think there's some general confusion regarding the rocm backend. This is a v0 only backend that will be deprecated once we have a working kernel for v1. A good template to look at would be the vllm/v1/attention/backends/flash_attn.py backend. Matching the flash_attn_varlen_func signature with this kernel would be a good goal.

rasmith commented:
> Very Exciting! I'll give it a try tomorrow morning. Thanks for the kernel!

@SageMoore Did this work for you?

rasmith commented:
> Hi, @rasmith. Are you planning to add block_table support to this kernel? If so, great! Feel free to either add that here or make a follow on PR. If not, I'd be happy to look into this, but we will need it in order to integrate this kernel into V1.

Would you be OK with a follow on PR? Do you have any other comments for the PR?

SageMoore commented:
@rasmith in this kernel's current state, it's only usable on v0 when the kv cache is empty. Are you seeing speedups over the previous implementation? If so, could you post your results to the PR? 

Assuming it does make things faster and we want to merge it, I'll give it a full review.  It would also be good to add in some unit tests since I suspect the CI integration testing for this pathway is somewhat lacking. We should still remove the backwards pass stuff as well.

rasmith committed:
resolve merge

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
unit tests work

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove bwd

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
reformatting

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
reformatting

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
reformatting

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
reformatting

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
reformatting

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
add unit tests

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
everything seems to work

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
codespell

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
fix incorrect function call

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
add spdx identifier

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

rasmith commented:
> @rasmith in this kernel's current state, it's only usable on v0 when the kv cache is empty. Are you seeing speedups over the previous implementation? If so, could you post your results to the PR?
> 
> Assuming it does make things faster and we want to merge it, I'll give it a full review. It would also be good to add in some unit tests since I suspect the CI integration testing for this pathway is somewhat lacking. We should still remove the backwards pass stuff as well.

@SageMoore 

I updated the PR with some times, the new kernel is a little faster than the older one.  I made sure the int8 and int8 kv cache stuff was there, but it's not hooked up yet, however, we can work on hooking it up later.

I also added the unit tests, although I'm not sure if they're taking up too much time.

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

rasmith committed:
update unit tests

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Try using newer fa kernel

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'main' into ransmith_triton_fav2_vsl

rasmith committed:
revert back to current triton fa

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
use older triton fa

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

rasmith committed:
Update unit tests, make it work with fp8 llama for ROCm

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
yapf

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
isort

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff/yapf

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove main from test file

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
mypy

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
mypy

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
isort

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
don't use fp8_out_scale parameter in abstract

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove cpa if statement

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
eight bit dtypes

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
eight bit dtype

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
eight bit dtype

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
init fp8_out_scale in layer.py

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove parameter

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
update autotune

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
merge main

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove variable redefinitions

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
add back in vscale float

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

LucasWilkinson commented on vllm/attention/backends/abstract.py:
```diff
@@ -237,6 +237,8 @@ class AttentionLayer(Protocol):
     _k_scale_float: float
     _v_scale_float: float
+    _prob_scale: torch.Tensor
+    _fp8_out_scale: torch.Tensor
```
nit: we should call this `_out_scale` to support future formats like fp4

ProExpertProg replied:
Also what's prob scale?

ilia-cher replied:
prob_scale is used to scale matrix multiplication result within FA loop

rasmith replied:
> nit: we should call this `_out_scale` to support future formats like fp4

Done

LucasWilkinson commented on vllm/model_executor/models/llama.py:
```diff
@@ -201,6 +209,8 @@ def forward(
         q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
         q, k = self.rotary_emb(positions, q, k)
+        self.attn._fp8_out_scale = (self.o_proj.input_scale
+                                    if self.attn_fp8_out else None)
```
I dont love that we have to do this in the model definition (the model definitions should be kept as simple as possible, also this requires every model definition to updated to support this)

We can potentially use a PyTorch compile pass to perform this fusion instead @ProExpertProg @tlrmchlsmth

ProExpertProg replied:
I agree we shouldn't be touching model definitions if possible. The only issue here might be that in V1 during piecewise compilation, the ops we're trying to fuse might exist in different subgraphs. But let's discuss this further before proceeding.

LucasWilkinson reviewed (commented):
left some comments, overall I think it looks good, but I do have concerns about touching the model definitions in-order to enable the fusion of fp8-quant onto the back of attention. Do you know how much of a performance hit there would be if we left that unfused? The reason I am asking is that in V1 we have torch.compile on by default and could potently use a torch.compile pass to do the fusion (@ProExpertProg is more aware of the challenges there) in a future PR

ProExpertProg commented on vllm/attention/backends/rocm_flash_attn.py:
```diff
@@ -699,6 +699,13 @@ def forward(
+                    full_scales = (
+                        layer._q_scale.item(), layer._k_scale.item(),
+                        layer._v_scale.item(), layer._prob_scale.item(),
+                        layer._fp8_out_scale.item()) if (
+                            layer._fp8_out_scale and layer._q_scale
+                            and layer._prob_scale
+                            and envs.VLLM_USE_ROCM_FP8_FLASH_ATTN) else None
```
This is hard to read, I'd extract to boolean

rasmith replied:
Done

ProExpertProg commented on vllm/envs.py:
```diff
@@ -13,6 +13,7 @@
     VLLM_NCCL_SO_PATH: Optional[str] = None
     LD_LIBRARY_PATH: Optional[str] = None
     VLLM_USE_TRITON_FLASH_ATTN: bool = False
+    VLLM_USE_ROCM_FP8_FLASH_ATTN: bool = False
```
+1 to both of these

rasmith replied:
used the kv cache dtype parameter

rasmith replied:
> can you please also provide accuracy results when using fp8

I posted PPL results for Llama-3.1-8B-Instruct-FP8-QKV-Pro vs Llama-3.1-8B-Instruct.  There are also unit tests that check accuracy.

gshtras replied:
Looking back here. Why did you decide to go with the kv-cache-dtype parameter? Isn't it a completely unrelated feature

ProExpertProg commented on vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w8a8_fp8.py:
```diff
@@ -23,6 +23,7 @@ class CompressedTensorsW8A8Fp8(CompressedTensorsScheme):
 
     def __init__(self, strategy: str, is_static_input_scheme: bool):
         self.strategy = strategy
+        self.out_dtype = torch.get_default_dtype()
```
I assume this depends on #14578?

ProExpertProg commented on vllm/model_executor/layers/quantization/utils/w8a8_utils.py:
```diff
@@ -172,8 +173,13 @@ def apply(
         # cutlass_scaled_mm supports per tensor/channel W and per tensor/token A
         if self.cutlass_fp8_supported:
+            assert input.dtype != current_platform.fp8_dtype(
+            ), "FP8 input to cutlass is not supported"
```
@LucasWilkinson that's not what this assert is

rasmith replied:
This file not changed in this PR anymore.

ProExpertProg reviewed (commented):
Some initial thoughts here. We should certainly not be touching model defionitions for this.

Honestly, I think we should just do fp8 integration in a separate PR. The kernel can stay as-is (with FP8 support), while the FP8 integration bits are moved to a new PR. That's because this kernel already contains many new features/improvements and it'll be easier to review and merge.

rasmith committed:
merge main

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove extra q_scale and is_navi

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
add booloean

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
use kv_cache_dtype

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove variable

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
rename _fp8_out_scale

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove fp8 integration and break into separate PR

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith commented:
> Some initial thoughts here. We should certainly not be touching model defionitions for this.
> 
> Honestly, I think we should just do fp8 integration in a separate PR. The kernel can stay as-is (with FP8 support), while the FP8 integration bits are moved to a new PR. That's because this kernel already contains many new features/improvements and it'll be easier to review and merge.

@SageMoore @LucasWilkinson @ProExpertProg  I split it up. Here is the PR for the fp8 integration: https://github.com/vllm-project/vllm/pull/15734

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+
+
+# TODO: This can probably optimized to have fewer lines of code.
+def get_strides_from_layout(q, k, v, o, metadata):
+    assert metadata.layout in SUPPORTED_LAYOUTS, "Got unsupported layout."
+    if metadata.layout == 'thd':
+        q_strides = (0, q.stride(1), q.stride(0), q.stride(2))
+        k_strides = (0, k.stride(1), k.stride(0), k.stride(2))
+        v_strides = (0, v.stride(1), v.stride(0), v.stride(2))
+        o_strides = (0, o.stride(1), o.stride(0), o.stride(2))
+    elif metadata.layout == 'bhsd':
+        q_strides = (q.stride(0), q.stride(1), q.stride(2), q.stride(3))
+        k_strides = (k.stride(0), k.stride(1), k.stride(2), k.stride(3))
+        v_strides = (v.stride(0), v.stride(1), v.stride(2), v.stride(3))
+        o_strides = (o.stride(0), o.stride(1), o.stride(2), o.stride(3))
+    elif metadata.layout == 'bshd':
+        q_strides = (q.stride(0), q.stride(2), q.stride(1), q.stride(3))
+        k_strides = (k.stride(0), k.stride(2), k.stride(1), k.stride(3))
+        v_strides = (v.stride(0), v.stride(2), v.stride(1), v.stride(3))
+        o_strides = (o.stride(0), o.stride(2), o.stride(1), o.stride(3))
+    return q_strides, k_strides, v_strides, o_strides
```
```suggestion
def get_strides_from_layout(q, k, v, o, metadata):
    STRIDE_PERMUTATIONS = {
        'thd': (None, 1, 0, 2),
        'bhsd': (0, 1, 2, 3),
        'bhsd': (0, 2, 1, 3),
    }
    assert metadata.layout in STRIDE_PERMUTATIONS, "Got unsupported layout."
    
    stride = lambda x, p: (0 if p is None else x.stride(p))
    strides = lambda x, ps: (stride(x, p) for p in ps)
    
    perm = STRIDE_PERMUTATIONS[metadata.layout]
    return (strides(x, perm) for x in [q, k, v, o])
```

rasmith replied:
It crashed and such, but I fixed it up and put it in.

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+# value   - [num_tokens, num_kv_heads, head_size
+# output  - [num_tokens, num_heads, head_size]
+def triton_attention(
+    q,
+    k,
+    v,
+    o,
+    cu_seqlens_q,
+    cu_seqlens_k,
+    max_seqlens_q,
+    max_seqlens_k,
+    causal=False,
+    sm_scale=1.0,
+    bias=None,
+    fp8_scales=None,
```
Could these get type annotations?

rasmith replied:
Done

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+            k = quantize_fp8(k, 1.0 / k_scale, eight_bit_dtype)
+            v = quantize_fp8(v, 1.0 / v_scale, eight_bit_dtype)
+
+        q_scale = torch.full((q.shape[1], 1, 1),
+                             q_scale,
+                             dtype=torch.float32,
+                             device=q.device)
+        k_scale = torch.full((k.shape[1], 1, 1),
+                             k_scale,
+                             dtype=torch.float32,
+                             device=k.device)
+        v_scale = torch.full((v.shape[1], 1, 1),
+                             v_scale,
+                             dtype=torch.float32,
+                             device=v.device)
+        p_scale = torch.full((q.shape[1], 1, 1),
+                             p_scale,
+                             dtype=torch.float32,
+                             device=q.device)
```
Why allocate these here instead of passing device tensors in the first place?

rasmith replied:
The tuple is a tuple of floats, however, the Triton kernel needs a tensor that has the right shape.

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+objects = {'tensors': {}}
+
+
+def quantize_fp8(x, scale, eight_bit_dtype):
+    float8_info = torch.finfo(eight_bit_dtype)
+    x *= scale
+    x = x.clamp(min=float8_info.min, max=float8_info.max).to(eight_bit_dtype)
+    return x
```
Can use `custom_ops.scaled_fp8_quant`

rasmith replied:
It doesn't work with uz fp8 datatype.

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -105,33 +302,36 @@ def _attn_fwd_inner(
+    QK_SCALE: tl.constexpr,
+    EIGHT_BIT_GEMM: tl.constexpr,
+    USE_P_SCALE: tl.constexpr,
+    EIGHT_BIT_KV: tl.constexpr,
```
Can you make sure all `bool` constexpr names start with `IS` or `USE` (or `DO`, whichever makes most sense)?

rasmith replied:
That looks to be the case, are there any that got missed?

rasmith replied:
Modified the names to whatever seemed to make the most sense

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -1,31 +1,158 @@
+3) Arbitrary head sizes
+4) Multi and grouped query attention
+5) Variable sequence lengths
+6) ALiBi and matrix bias
```
Add fp8 support to the list?

rasmith replied:
Done

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -163,156 +386,259 @@ def _attn_fwd_inner(
+
+        if EIGHT_BIT_GEMM:
+            if USE_P_SCALE:
+                p = (p * p_scale).to(EIGHT_BIT_DTYPE)
```
Should we also clamp here? Or better yet call just call a `quant` subroutine?

rasmith replied:
Done

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -163,156 +386,259 @@ def _attn_fwd_inner(
+        else:
+            if EIGHT_BIT_KV:
+                v = (v * v_descale).to(p.type.element_ty)
+            acc += tl.dot(p.to(v.type.element_ty), v)
```
Why is `p` converted here, isn't it always 16-bit here? I guess a redundant cast is better than a missing one

rasmith replied:
It is to ensure that p and v are the same type so triton can emit IR instead of crashing (logically it could promote, but it doesn't seem to do that right now).

ProExpertProg commented on tests/kernels/test_triton_flash_attention.py:
```diff
@@ -0,0 +1,607 @@
+FP8_MAX = float8_info.max
+
+
+def get_shape_from_layout(q, k, metadata):
```
Duplicate?

rasmith replied:
De-duped

ProExpertProg commented on tests/kernels/test_triton_flash_attention.py:
```diff
@@ -0,0 +1,607 @@
+                                                       compute_alibi_tensor,
+                                                       triton_attention_rocm)
+
+FP8_DTYPE_TORCH = torch.float8_e4m3fnuz
+
+float8_info = torch.finfo(FP8_DTYPE_TORCH)
+FP8_MIN = float8_info.min
+FP8_MAX = float8_info.max
```
Generally we use the `torch.float8_e4m3fn` max/min here for compatibility with other devices. Are you sure we want to do the full `uz` range here?

rasmith replied:
Just using current_platform.fp8_dtype() now

ProExpertProg replied:
No that's wlnot what I'm referring to. I mean that even when the type is fnuz, we still use the max of fn (check out quant_adjusted_max in the CUDA/hip kernels).

rasmith replied:
I searched for quant_adjusted_max and can't find it.

rasmith replied:
I searched for adjusted_max and also can't find it.

ProExpertProg replied:
It's called `quant_type_max` and `quant_type_max_v` - sorry for misleading you with the name

rasmith replied:
Messaged on slack

ProExpertProg commented on tests/kernels/test_triton_flash_attention.py:
```diff
@@ -0,0 +1,607 @@
+    return batch, nheads_q, nheads_k, head_size
+
+
+def quantize_fp8(tensor: torch.Tensor,
+                 dim) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    max_vals = tensor.abs().amax(
+        dim=[i for i in range(tensor.dim()) if i != dim], keepdim=True)
+
+    # Avoid division by zero
+    max_vals[max_vals == 0] = 1e-8
+
+    # Compute scale factors for each channel
+    scale = (FP8_MAX / max_vals).clamp(1e-12)
+
+    # Quantize the tensor
+    tensor = tensor * scale
+    tensor.clamp_(FP8_MIN, FP8_MAX)
+    tensor_quantized = tensor.to(FP8_DTYPE_TORCH)
+
+    return tensor_quantized, scale, 1 / scale
```
Could use `custom_ops.scaled_fp8_quant`?

rasmith replied:
It doesn't work for nuz types.

ProExpertProg replied:
If so could you just extend the support for it? We'd rather not add new kernels if we can extend a current kernel

rasmith replied:
I think someone at AMD tried, and could not get it to work, then I think I tried briefly on a related project, and couldn't get it to work, and then Jeff Daily, our PyTorch guy did some work on these things, and the function still crashes with a 'not supported' error for the torch.float8_e4m3fnuz data type.  I did manage to get it to run on our downstream ROCm vllm, but it would segfault.

ProExpertProg replied:
I'll take a look on monday

rasmith replied:
I think there are other scaled_XXX functions that have the same problem too, some of them may have been fixed by Jeff Daily, the PR was this one: https://github.com/vllm-project/vllm/pull/14245

rasmith replied:
OK, I looked at this again, and I tried running scaled_fp8_quant and got this:

>>> ops.scaled_fp8_quant(tensor)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/vllm-upstream-moe/vllm/_custom_ops.py", line 906, in scaled_fp8_quant
    assert (input.ndim == 2)
    

I think this function is quantizing 4-dimensional tensors taking the max from a given axis, e.g. dim=1 or dim=2.
The reason, I think, is so the kernel can still handle the bshd and bhsd layouts, which makes it possible for the kernel to take the same layouts as the upstream triton_flash_attention in Tri Dao's repository.

So, it looks like scaled_fp8_quant can't 1) quantize using the max from a given axis, and 2) quantize a 3 or 4 dimensional tensor.

Is there a function in vLLM that can do this?

ProExpertProg replied:
You could do something similar to what `Fp8Linear` [does](https://github.com/vllm-project/vllm/blob/4d0ec37267afaf988e32174ebc31f24268076491/vllm/model_executor/layers/quantization/utils/w8a8_utils.py#L178):
```
input_2d = input.view(-1, input.shape[-1])
```

rasmith replied:
Still need to take the max from the appropriate dimension (in this case dim=1 or dim=2).

rasmith replied:
It might not matter too much though, I could just reshape it, quantize, and just use whatever scaled_fp8_quant picks.

ProExpertProg commented on tests/kernels/test_triton_flash_attention.py:
```diff
@@ -0,0 +1,607 @@
+                                       input_metadata)
+
+    # Compute scores
+    scores = torch.einsum('bhqd,bhkd->bhqk', q,
+                          k_dequantized).float() * input_metadata.sm_scale
+
+    if causal:
+        mask = torch.tril(torch.ones(N_CTX_Q, N_CTX_K, device="cuda"),
+                          diagonal=N_CTX_K - N_CTX_Q)
+        scores[:, :, mask == 0] = float("-inf")
+
+    p = torch.softmax(scores, dim=-1)
+    ref_out = torch.einsum('bhqk,bhkd->bhqd', p.half(),
+                           v_dequantized).to(torch.float16)
```
Could you extract the reference implementations from all the tests? Perhaps into a `ReferenceAttention` class?

rasmith replied:
Done

ProExpertProg reviewed (commented):
Thanks for this kernel and sorry the review took so long. I mostly looked at the FP8 bits and it looks pretty good, but I had a bunch of small code organization comments.

I think we should lift the `quantize_fp8` calls out of the triton kernel and into the integration code, so `q`/`k`/`v` values arrive quantized to the kernel. That way we can also support dynamic quantization more easily.

I also think all the `eight_bit` stuff is verbose. Generally in the codebase we use the name `fp8_dtype` and not `eight_bit_type`. If this kernel is supposed to generalize to `int8` we can revisit that later and we'd probably still call the type `quant_type`.

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+        eight_bit_dtype = current_platform.fp8_dtype()
+
+        (q_scale, k_scale, v_scale, p_scale, o_scale) = fp8_scales
+        if q.dtype != eight_bit_dtype:
```
I know I asked you to move this out but wherever you move it to, I'd probably at least assert `q`/`k`/`v` all have the same type

rasmith replied:
Moved some of it out to functions.

ProExpertProg reviewed (commented):
Had a few more comments regarding output scaling. I think we should try to keep the features of the kernel as orthogonal as possible, and even if we don't support all combinations of features, it should be very clear what scenarios we support and what flags/parameters need to be passed in each case.

EDIT: To clarify, I'm referring to the variety of different tensors and whether they're in fp8 or not

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -1,31 +1,158 @@
+SUPPORTED_LAYOUTS = ['thd', 'bhsd', 'bshd']
+
+default_eight_bit_dtype_triton = tl.float8e4b8
+default_eight_bit_dtype_torch = torch.float8_e4m3fnuz
```
It's just a default type, the actual type will get picked up during execution, e.g. it will get set in the triton_attention function to be the actual type encountered in the tensor .  I'd like to avoid importing too much stuff here.

ProExpertProg replied:
I don't think I understand why avoiding the import is not worth it here

rasmith replied:
I added it in, but I've been bitten by vllm circular imports quite a few times.

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -61,39 +188,109 @@ def dropout_mask(philox_seed, philox_offset, dropout_p, m, n, stride):
-        tensor = tl.load(block_ptr, boundary_check=(0, ), padding_option=pad)
-    elif second:
-        tensor = tl.load(block_ptr, boundary_check=(1, ), padding_option=pad)
+def load_fn(ptrs, offset_first, offset_second, boundary_first,
```
Done

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+                    block_min = block_max
+                    block_max = n_blocks * BLOCK_N
+
+                tl.debug_barrier()
```
I believe so, it's also present in upstream FA kernels, example, https://github.com/Dao-AILab/flash-attention/blob/2dd8078adc1d9b74e315ee99718c0dea0de8eeb6/flash_attn/flash_attn_triton.py#L487.

Seems there are certain race conditions that do occur without it.  I can ask about it if you like.

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+                causal_start_idx = seqlen_q - seqlen_k
+                if EIGHT_BIT and not EIGHT_BIT_KV:  # noqa: SIM102
+                    if o_descale is not None:
+                        acc *= o_descale
+                        acc = tl.clamp(acc, FP8_MIN, FP8_MAX)
```
What is quant subroutine?

rasmith replied:
Also, do to the output scaling now happening in RowParallelLinear in the other PR, this ended up being removed anyways.

rasmith replied:
OK, used something like this.

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+                start_m_idx = start_m * BLOCK_M
+                causal_start_idx = seqlen_q - seqlen_k
+                if EIGHT_BIT and not EIGHT_BIT_KV:  # noqa: SIM102
+                    if o_descale is not None:
```
This was removed.

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+            q_descale, k_descale, p_scale, p_descale, v_descale, o_scale = (
+                metadata.q_descale, metadata.k_descale, metadata.p_scale,
+                metadata.p_descale, metadata.v_descale, metadata.o_scale)
+            o_descale = 1.0 / o_scale if o_scale is not None else None
```
This was removed.

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -163,156 +386,259 @@ def _attn_fwd_inner(
 
-@triton.autotune(
-    configs=[
+def is_hip():
+    return triton.runtime.driver.active.get_current_target().backend == "hip"
```
Done

rasmith commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,699 @@ def attn_fwd(
+    return q_strides, k_strides, v_strides, o_strides
 
 
 class _attention(torch.autograd.Function):
 
     @staticmethod
```
I think it was inherited from the Tri Dao's upstream FA flash-attention repo (which this kernel is based on).  They have backward passes and stuff. It does create a node in the computation graph (maybe that's useful), but otherwise wasn't sure how necessary it was.

rasmith committed:
clean up unit tests and triton kernel

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
ruff

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -1071,10 +1069,6 @@ def attn_fwd(
                 end_m_idx = (start_m + 1) * BLOCK_M
                 start_m_idx = start_m * BLOCK_M
                 causal_start_idx = seqlen_q - seqlen_k
-                if EIGHT_BIT and not EIGHT_BIT_KV:  # noqa: SIM102
```
We should keep this support in the kernel, we just don't want to integrate it through the model definition. But the kernel should still support the fused quantization of the output

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -667,7 +668,6 @@ def attn_fwd(
     K_descale,
     P_scale,
     P_descale,
-    o_descale,
```
No need to remove this parameter

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -1146,25 +1140,20 @@ def get_shape_from_layout(q, k, metadata):
-        o_strides = (o.stride(0), o.stride(2), o.stride(1), o.stride(3))
-    return q_strides, k_strides, v_strides, o_strides
+
+    STRIDE_PERMUTATIONS = {
```
I think `SUPPORTED_LAYOUTS` is just the keys of `STRIDE_PERMUTATIONS` right? So maybe extract `STRIDE_PERMUTATIONS` and get rid of `SUPPORTED_LAYOUTS` (or define them explicitly as the keys).

rasmith replied:
Not sure what real improvement it is:

I could just do 
  
  assert metadata.layout in STRIDE_PERMUTATIONS.keys()
  
 but SUPPORTED_LAYOUTS is still needed elsewhere and
  
  assert layout in SUPPORTED_LAYOUTS
  
 is more readable than using
  
  assert layout in STRIDE_PERMUTATIONS.keys()
  
  anywhere I see it.

ProExpertProg reviewed (commented):
A few more comments. Most importantly, don't remove support for fused quantization from the Triton kernel, we will invoke that codepath with a torch.compile transformation. Just removing it from the other (integration) PR is enough.

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -352,470 +693,705 @@ def attn_fwd(
+
+def prepare_scales_for_triton(q, k, v, fp8_scales):
+    (q_scale, k_scale, v_scale, p_scale) = fp8_scales
+    q_scale = torch.full((q.shape[1], ),
+                         q_scale,
+                         dtype=torch.float32,
+                         device=q.device)
+    k_scale = torch.full((k.shape[1], ),
+                         k_scale,
+                         dtype=torch.float32,
+                         device=k.device)
+    v_scale = torch.full((v.shape[1], ),
+                         v_scale,
+                         dtype=torch.float32,
+                         device=v.device)
+    p_scale = torch.full((q.shape[1], ),
+                         p_scale,
+                         dtype=torch.float32,
+                         device=q.device)
```
This is manually broadcasting from a scalar value, right? Why not just broadcast implicitly inside Triton?

rasmith replied:
I forgot, the kernel does load a tensor, but it this tensor has shape of [1,48,1,1], but it loads them one at a time, and then it accommodate GQA/MQA.  So I think I can add a flag to tell the kernel whether it has a singular value or not in the tensor and that might work fine.

rasmith committed:
output fp16 from triton FA, update test for o_scale, accept only tensor, update some scaling

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
overwrote wrong file

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'main' into ransmith_triton_fav2_vsl

rasmith committed:
add is_rocm_rdna and is_rocm_cdna methods

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

ProExpertProg commented on vllm/platforms/interface.py:
```diff
@@ -181,6 +187,22 @@ def has_device_capability(
 
+    @classmethod
+    def has_rdna_target(cls):
+        import triton
+        ROCM_RDNA_TARGETS = [
+            "gfx1030", "gfx1100", "gfx1101", "gfx1102", "gfx1200", "gfx1201"
+        ]
+        return triton.runtime.driver.active.get_current_target(
+        ).arch in ROCM_RDNA_TARGETS
```
We should return `False` here and override in the ROCm platform

rasmith replied:
Done

rasmith committed:
move has_rdna_target/has_cdna_target impl to rocm.py

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

ProExpertProg commented on tests/kernels/test_triton_flash_attention.py:
```diff
@@ -0,0 +1,585 @@
+FP8_DTYPE_TORCH = current_platform.fp8_dtype()
+
+float8_info = torch.finfo(FP8_DTYPE_TORCH)
+FP8_MIN = float8_info.min
+FP8_MAX = float8_info.max
```
From my experience, if you use `tl.constexpr`, you can use these global constants in kernels directly:
```suggestion
FP8_MIN: tl.constexpr = float8_info.min
FP8_MAX: tl.constexpr = float8_info.max
```

rasmith replied:
Done

rasmith committed:
use tl.constexpr in kernel instead of passing

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
use scaled_fp8_quant and simplify unit tests more

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
add subroutine to quantize fp8 triton tensors

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

BowenBao commented:
Hi @rasmith thanks for the PR! Question on this part
>This kernel supports int8 KV cache, and also the following (forward only):

Is it a typo or do you mean the kernel also supports int8 kv cache?

rasmith commented:
> Hi @rasmith thanks for the PR! Question on this part
> 
> > This kernel supports int8 KV cache, and also the following (forward only):
> 
> Is it a typo that should be fp8 instead or do you mean the kernel also supports int8 kv cache?

Well, eventually, I may be adding int8 implementations for int8 kv-cache, depending on priority, so this does add "support" for it, but more work will need to be done to actually make it happen.

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -36,6 +36,9 @@
 default_float8_info = torch.finfo(default_eight_bit_dtype_torch)
 
+DEFAULT_FP8_MIN: triton.language.constexpr = default_float8_info.min
+DEFAULT_FP8_MAX: triton.language.constexpr = default_float8_info.max
```
Why not just call these `FP8_MIN`/`FP8_MAX` and avoid the copy below?

ProExpertProg commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -1321,9 +1319,7 @@ def forward(ctx, q, k, v, o, metadata: MetaData):
-                       EIGHT_BIT_DTYPE=metadata.eight_bit_dtype_triton,
-                       FP8_MIN=float8_info.min,
-                       FP8_MAX=float8_info.max)
+                       EIGHT_BIT_DTYPE=metadata.eight_bit_dtype_triton)
```
Call this `QUANT_DTYPE`?

rasmith replied:
Done

ProExpertProg reviewed (commented):
I still think `EIGHT_BIT` should become either `QUANT` or `FP8`, depending on whether it'll support other quant types in the future or not. We don't really use `eight_bit` as an identifier in the vLLM codebase (that I'm aware of)

ProExpertProg commented on vllm/platforms/interface.py:
```diff
@@ -120,6 +120,12 @@ def is_cuda(self) -> bool:
     def is_rocm(self) -> bool:
         return self._enum == PlatformEnum.ROCM
 
+    def is_rocm_cdna(self) -> bool:
+        return self.is_rocm() and self.has_cdna_target()
+
+    def is_rocm_rdna(self) -> bool:
+        return self.is_rocm() and self.has_rdna_target()
```
I don't think both methods are required, unless RDNA/CDNA has equivalents on other platforms? Just implement one of `is_rocm_cdna`/`has_cdna_target` and override it on ROCm.

rasmith replied:
Done

rasmith committed:
remove rocm_cdna and rename bool vars

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove is_rdna_target

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
persistent->is_persistent

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

SageMoore commented on vllm/platforms/interface.py:
```diff
@@ -120,6 +120,9 @@ def is_cuda(self) -> bool:
     def is_rocm(self) -> bool:
         return self._enum == PlatformEnum.ROCM
 
+    def is_rocm_cdna(self) -> bool:
```
Nit: instead of putting this in the platform class, can you add an is_rocm_cdna function to the triton_flash_attention file? I generally try to avoid adding methods to the Platform class unless they are relevant for all platforms. For example, we don't have a is_cuda_h100 method in this file :).

rasmith replied:
Seems fine to me either way, but @ProExpertProg wanted it here.

ProExpertProg replied:
I think it's fair we don't want it on all platforms but I think kernels and device features shouldn't mix, and usually we put features in the platform class. I could be convinced otherwise

rasmith replied:
@ProExpertProg @SageMoore 

I could add something like this to Platform:

    @classmethod
    def query_device(cls, query):
        if hasattr(cls, query):
            fn = getattr(cls, query)
            if callable(fn):
                return fn()
        return False

And then it's possible to do:

current_platform.query_device("has_cdna_target")

ProExpertProg replied:
I like the approach but if you make `query_platform` into `__getattr__` (with some changes) then it could return `False` for any method that starts with `has_` and your line becomes `current_platform.has_cdna_target()`.

rasmith replied:
moved the cdna query back into the kernel

SageMoore commented on tests/kernels/test_triton_flash_attention.py:
```diff
@@ -0,0 +1,551 @@
+
+
+@pytest.mark.parametrize('Z, HQ, HK, N_CTX_Q, N_CTX_K, D_HEAD', [
+    (4, 48, 12, 1, 1, 64),
```
Nit: Let's add a few additional batches sizes here. 1, 16, and 64 should be a good mix. Same idea for the other tests.

ProExpertProg replied:
Also please add tests for other layouts here, running into issues after fusion

rasmith replied:
@ProExpertProg @SageMoore  It already takes 3-4 minutes to run the tests.  Adding in other layouts plus the extra batch sizes might make this take almost 30 min.  Any suggestions?

ProExpertProg replied:
You could linearize the test cases so you're not checking all combinations, just some with each layout, and with one layout you check all of them

rasmith replied:
@ProExpertProg The varlen tests test the thd layout: test_op_varlen_fwd and test_op_varlen_mqa_fwd.  I'll add a note for these tests.

@SageMoore I added some additional batch sizes to test_op_fwd and test_op_varlen_fwd

I removed the bhsd layout from the test_op_fwd test since bhsd is tested in other tests.  Runtime is roughly 3 min.

SageMoore commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -1,31 +1,174 @@
+                                      != torch.float8_e4m3fnuz else 224.0)
+
+
+class MetaData:
```
I have a general question about this `MetaData` class. How would you feel about expanding the `__init__` method to initialize all of these instance variables instead of relying on the caller to set everything manually. It looks like you initialize this data structure in every call to the kernel, meaning that all of the data it needs is present when it's allocated. The reason I'm suggesting this is because you could have a soft "guarantee" that this class is always in a valid state if you pass everything in at `__init__` time .

rasmith replied:
Well, I could do something like this:

```
def __init__(self,
                  sm_scale = 1.0,
                  is_varlen = False,
                  cu_seqlens_q = None,
                  cu_seqlens_k = None,
                  # more params here
):
    if is_varlen:
        self.set_varlen_params(cu_seqlens_q, cu_seqlens_k)
     # etc. for rest of the helper methods / params
```

@ProExpertProg @SageMoore  

In terms of being in a good state, the object is still in a tentatively good state after calling each helper method and is also in a tentatively good state after calling __init__ method with the extra params.

I think it can add some convenience, although convenience is a matter of taste in some regards, and possible kwarg usage.

What do you think?

ProExpertProg replied:
I think that's fine. You could also add a @classmethod for each init scenario and call it like `metadata = Metadata.varlen(...)` and whatever the other case is

rasmith replied:
Done

SageMoore commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -335,487 +668,736 @@ def attn_fwd(
-        )
+            USE_BIAS=metadata.bias is not None,
+            USE_ALIBI=metadata.alibi_slopes is not None,
+            SHOULD_ENABLE_DROPOUT=metadata.dropout_p > 0.0,
```
Can we simplify the code a bit by removing all of the dropout stuff?

rasmith replied:
Done

SageMoore reviewed (commented):
Thanks for the contribution @rasmith! I've left you an initial round of comments. It would be great to get lm_eval results posted here for the fp8 llama models that you used to test. That coupled with the unit test should give us reasonable confidence that everything is working well.

rasmith committed:
remove dropout

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove unnecessary params

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
move the cdna check back into triton kernel

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove bhsd from some tests and add more layout choices and annotated the varlen tests for thd layout

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
refactor metadata __init__ and use in tests

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

tlrmchlsmth commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -335,487 +668,736 @@ def attn_fwd(
+                            p_descale_ptrs = p_descale_ptr + off_h_q
+
+                if USE_BIAS:
+                    # Note: might get large enough to overflow on some configs
+                    bias_offset = off_h_q * stride_bh
```
Does this make sense?
```suggestion
                    bias_offset = off_h_q * tl.cast(stride_bh, t.int64)
```

rasmith replied:
Done

tlrmchlsmth commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -335,487 +668,736 @@ def attn_fwd(
+                    bias_ptrs = None
+
+                if USE_ALIBI:
+                    a_offset = off_z * stride_az + off_h_q * stride_ah
```
similar, does it make sense to do:
```suggestion
                    a_offset = off_z * tl.cast(stride_az, tl.int64) + off_h_q * tl.cast(stride_ah, tl.int64)
```

rasmith replied:
Done

tlrmchlsmth commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -335,487 +668,736 @@ def attn_fwd(
+                                       off_h_q * seqlen_q * seqlen_k)
+                    encoded_sm_ptrs = (encoded_sm_base +
+                                       offs_m[:, None] * seqlen_k +
+                                       offs_n[None, :])
```
There's some spots here that look like they might overflow an int32 as well:
`off_hz = off_z * HQ + off_h_q` and `off_hz * seqlen_q * seqlen_k`
and
`off_h_q * seqlen_q * seqlen_k` and `offs_m[:, None] * seqlen_k`

rasmith replied:
Done

tlrmchlsmth commented on vllm/attention/ops/triton_flash_attention.py:
```diff
@@ -335,487 +668,736 @@ def attn_fwd(
+                                                 != BLOCK_DMODEL)
+
+                # Compute pointers for all the tensors used in this kernel.
+                q_offset = (Q + off_z * stride_qz + off_h_q * stride_qh +
+                            cu_seqlens_q_start * stride_qm)
+                q_ptrs = (q_offset + offs_m[:, None] * stride_qm +
+                          offs_d[None, :] * stride_qk)
+                k_offset = (K + off_z * stride_kz + off_h_k * stride_kh +
+                            cu_seqlens_k_start * stride_kn)
+                k_ptrs = (k_offset + offs_d[:, None] * stride_kk +
+                          offs_n[None, :] * stride_kn)
+                v_offset = (V + off_z * stride_vz + off_h_k * stride_vh +
+                            cu_seqlens_k_start * stride_vk)
+                v_ptrs = (v_offset + offs_n[:, None] * stride_vk +
+                          offs_d[None, :] * stride_vn)
```
potential overflow concerns here as well?

rasmith replied:
Here and in other places, I checked for possible overflow cases and casted to int64.

tlrmchlsmth reviewed (commented):
I noted some spots where strides are being multiplied by offsets, and integer overflows seemed like they could be likely. (I think triton uses int32 by default?). Could you check those spots?

rasmith committed:
remove has_cdna_target from interface.py

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith commented on vllm/platforms/interface.py:
```diff
@@ -181,6 +181,10 @@ def has_device_capability(
 
+    @classmethod
+    def has_cdna_target(cls):
+        return False
```
Meant to remove this, so it's removed.

rasmith committed:
handle cases where integer overflow could occur

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove comment

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

tlrmchlsmth reviewed (approved):
This LGTM now, thank you!

ProExpertProg commented:
@mgoin could you manually kick off the docker build with extended timeout?

rasmith committed:
revert to see if segv goes away

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
try avoiding overflow while also avoiding crashing nvidia triton compiler

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
make bools constexpr so Triton compiler doesn't crash

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
remove persistent

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
adjust tolerances for unit tests

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
requires grad and adjust o_scale test param

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

rasmith committed:
use triton.language.constexpr correctly and add check for cuda 9.0 for Triton fp8

Signed-off-by: Randall Smith <Randall.Smith@amd.com>

rasmith committed:
Merge branch 'ransmith_triton_fav2_vsl' of github.com:rasmith/vllm into ransmith_triton_fav2_vsl

rasmith committed:
Merge branch 'main' into ransmith_triton_fav2_vsl

rasmith committed:
Merge branch 'main' into ransmith_triton_fav2_vsl

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl

rasmith committed:
Merge branch 'vllm-project:main' into ransmith_triton_fav2_vsl
